scheduler/utilization_based_host_allocator.go

```diff
@@ -0,0 +1,248 @@
+package scheduler
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/evergreen-ci/evergreen/model"
+	"github.com/evergreen-ci/evergreen/model/distro"
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/evergreen-ci/evergreen/model/task"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+)
+
+func UtilizationBasedHostAllocator(ctx context.Context, hostAllocatorData HostAllocatorData) (map[string]int, error) {
+	count := 0
+	var distroName string
+	newHostsNeeded := make(map[string]int)
+
+	// once the input format changes this loop can be removed - should only have 1 distro
+	for name, d := range hostAllocatorData.distros {
+		distroName = name
+		if count > 0 {
+			msg := "more than 1 distro sent to UtilizationBasedHostAllocator"
+			grip.Emergency(message.Fields{
+				"message": msg,
+				"distros": hostAllocatorData.distros,
+			})
+			return nil, errors.New(msg)
+		}
+
+		// actual calculation logic is here
+		newHosts, err := numNewHosts(ctx, d, hostAllocatorData.taskQueueItems[name],
+			hostAllocatorData.existingDistroHosts[name], hostAllocatorData.freeHostFraction)
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "error calculating hosts for distro %s", name)
+		}
+		newHostsNeeded[name] = newHosts
+		count++
+	}
+
+	grip.Info(message.Fields{
+		"runner":        RunnerName,
+		"num_new_hosts": newHostsNeeded[distroName],
+		"message":       "requsting new hosts",
+	})
+
+	return newHostsNeeded, nil
+}
+
+// Calculate the number of hosts needed by taking the total task scheduled task time
+// and dividing it by the target duration. Request however many hosts are needed to
+// achieve that minus the number of free hosts
+func numNewHosts(ctx context.Context, d distro.Distro, taskQueue []model.TaskQueueItem,
```

Comment:
I'm sure that we'll need a better function name for this.

Reply:
Renamed, hopefully so that it's clearer